(CDAP-16730) Set correct -xmx for pods in k8s to run programs

wyzhang opened the pull request:
Bug:
We have introduced a multiplier to determine the actual CPU and RAM
to request for creating the pod to run a program. But we fail to
set the correct -xmx based on the RAM requested for the pod.

This change
- Set memory requested to be memory * multiplier
- Set xmx to be memory * multiplier * 0.8
where memory is the requested memroy by the program, we may allocate
more or less depending on the setting of multplier which depends on
the k8s cluster size.

chtyim commented on cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/runtime/KubeTwillPreparer.java:
```diff
@@ -87,6 +87,7 @@
   private static final String CPU_MULTIPLIER = "master.environment.k8s.container.cpu.multiplier";
   private static final String MEMORY_MULTIPLIER = "master.environment.k8s.container.memory.multiplier";
   private static final String DEFAULT_MULTIPLIER = "1.0";
+  private static final double MEMORY_MAX_HEAP_SIZE_FRACTION = 0.8;
```
We have two keys in the cdap-site.xml to control non-heap memory, `twill.java.heap.memory.ratio` and `twill.java.reserved.memory.mb`. You can get those through the `MasterEnvironmentContext.getConfigurations()` method. You can refer to the `org.apache.twill.internal.utils.Resources.computeMaxHeapSize` method on how it computes the xmx.

wyzhang replied:
Done. Added twill as dependency and uses the util function directly.

chtyim commented on cdap-kubernetes/pom.xml:
```diff
@@ -76,6 +76,10 @@
     </dependency>
+    <dependency>
+      <groupId>org.apache.twill</groupId>
+      <artifactId>twill-core</artifactId>
```
twill-core has quite a lot of dependencies. If you are only using one class, better to exclude all the dependencies. The Resources class in Twill has no dependency on non JDK classes

wyzhang replied:
Done. Excluded all dependencies of twill-core

wyzhang committed:
(CDAP-16730) Set correct -xmx for pods in k8s to run programs

Bug:
We have introduced a multiplier to determine the actual CPU and RAM
to request for creating the pod to run a program. But we fail to
set the correct -xmx based on the RAM requested for the pod.

This change
- Set memory requested to be memory * multiplier
- Set xmx using org.apache.twill.internal.utils.Resources.computeMaxHeapSiz
  based on the memory request size
